src/rdkafka.h

```diff
@@ -4202,6 +4293,975 @@ rd_kafka_interceptor_add_on_request_sent (
+rd_kafka_NewTopic_destroy_array (rd_kafka_NewTopic_t **new_topics,
+                                 size_t new_topic_cnt);
+
+
+/**
+ * @brief Set the replica (broker) assignment for \p partition to the
+ *        replica set in \p broker_ids (of \p broker_id_cnt elements).
+ *
+ * @remark When this method is used, rd_kafka_NewTopic_new() must have
+ *         been called with a \c replication_factor of -1.
+ *
+ * @remark An application must either set the replica assignment for
+ *         all new partitions, or none.
+ *
+ * @remark If called, this function must be called consecutively for each
+ *         partition, starting at 0.
+ *
+ * @remark Use rd_kafka_metadata() to retrieve the list of brokers
+ *         in the cluster.
+ *
+ * @returns RD_KAFKA_RESP_ERR_NO_ERROR on success, or an error code
+ *          if the arguments were invalid.
+ *
+ * @sa rd_kafka_AdminOptions_set_validate_only()
+ */
+RD_EXPORT rd_kafka_resp_err_t
+rd_kafka_NewTopic_set_replica_assignment (rd_kafka_NewTopic_t *new_topic,
+                                          int32_t partition,
+                                          int32_t *broker_ids,
+                                          size_t broker_id_cnt,
+                                          char *errstr, size_t errstr_size);
+
+/**
+ * @brief Add a (broker-side) topic configuration name/value pair to
+ *        the NewTopic object.
+ *
+ * @remark The name and value are not validated by the client, the validation
+ *         takes place on the broker.
+ *
+ * @returns RD_KAFKA_RESP_ERR_NO_ERROR on success, or an error code
+ *          if the arguments were invalid.
+ *
+ * @sa rd_kafka_AdminOptions_set_validate_only()
+ * @sa http://kafka.apache.org/documentation.html#topicconfigs
+ */
+RD_EXPORT rd_kafka_resp_err_t
+rd_kafka_NewTopic_add_config (rd_kafka_NewTopic_t *new_topic,
+                              const char *name, const char *value);
+
+
+/**
+ * @brief Create topics in cluster as specified by the \p new_topics
+ *        array of size \p new_topic_cnt elements.
+ *
+ * @param new_topics Array of new topics to create.
+ * @param new_topic_cnt Number of elements in \p new_topics array.
+ * @param options Optional admin options, or NULL for defaults.
+ * @param rkqu Queue to emit result on.
+ *
+ * Supported admin options:
+ *  - rd_kafka_AdminOptions_set_validate_only() - default false
+ *  - rd_kafka_AdminOptions_set_operation_timeout() - default 0
+ *  - rd_kafka_AdminOptions_set_timeout() - default socket.timeout.ms
+ *
+ * @remark The result event type emitted on the supplied queue is of type
+ *         \c RD_KAFKA_EVENT_CREATETOPICS_RESULT
+ */
+RD_EXPORT void
+rd_kafka_CreateTopics (rd_kafka_t *rk,
+                       rd_kafka_NewTopic_t **new_topics,
+                       size_t new_topic_cnt,
+                       const rd_kafka_AdminOptions_t *options,
+                       rd_kafka_queue_t *rkqu);
+
+
+/**
+ * @brief CreateTopics result type and methods
```

Comment:
are you expecting binding libraries to use a common queue for all concurrent requests or a separate queue per request? If there are substantial advantages to using a single queue (that I'm not aware of), I think we really want an opaque pointer that can be passed into `rd_kafka_CreateTopics` which is then made available via `rd_kafka_CreateTopics_result_t`, otherwise we need to match up requests with results by looking at topic names.

I expect queue per request is fine, and I'm going to set things up like that in the C# binding unless otherwise informed... this approach also has the advantage of avoiding the need to pin memory in a garbage collected language.

Reply:
If you want to provide a blocking interface then you would typically use a temporary per API request, but if you want it async and poll the results from a single place you would use a shared queue, be it a specific one of for admin results or the main queue.

See rd_kafka_AdminOptions_set_opaque()